### DISCUSSION TITLE
Override Argument

### DISCUSSION THREAD

**Discussion by sschmocker** [Romanitho/Winget-Install#20] [Q&A] [answered]
I cant seem to get it to work, am I doing something wrong?
powershell.exe -Executionpolicy Bypass -File winget-install.ps1 -AppIDs "Citrix.Workspace --override "/silent /noreboot /includeSSON /forceinstall""

**Comment by Romanitho** [comment]
This is due to your quote. Try to escape
What you coud try as well:
`powershell.exe -Executionpolicy Bypass -Command .\winget-install.ps1 -AppIDs 'Citrix.Workspace --override "/silent /noreboot /includeSSON /forceinstall"'`

**Reply by sschmocker** [reply]
> This is due to your quote. Try to escape What you coud try as well: `powershell.exe -Executionpolicy Bypass -Command .\winget-install.ps1 -AppIDs 'Citrix.Workspace --override "/silent /noreboot /includeSSON /forceinstall"'`

I tried that, but that also Fails.

**Reply by Romanitho** [reply]
try escape the quote with \ like this :
`powershell.exe -Executionpolicy Bypass -File winget-install.ps1 -AppIDs "Citrix.Workspace --override \"/silent /noreboot /includeSSON /forceinstall\""`

**Comment by Romanitho** [comment]
@KnifMelti nice example here to troubleshoot

**Reply by KnifMelti** [reply]
Ah, Workspace.
We use it too... (who doesn't ;))

**Reply by KnifMelti** [reply]
These are my findings (using Start-Process **not** cmd.exe - not yet available in source - @Romanitho could you test with the current solution, please?) and **$AppArgs** running in System context:

```
#Install App
Write-Log "Installing $AppID..." "Yellow"
$Command = "install --id $AppID --silent --accept-package-agreements --accept-source-agreements $AppArgs"
Start-Process -FilePath "$winget" -ArgumentList "$Command" -NoNewWindow -Wait
```

Run:
`powershell.exe -Executionpolicy Bypass -File winget-install.ps1 -AppIDs "Citrix.Workspace --override \"/silent /noreboot /includeSSON /forceinstall\""`

Result (in C:\Windows\Temp\WinGet\WinGet*****.log):
`2022-04-12 21:19:12.171 [CORE] Command line Args: "C:\Program Files\WindowsApps\Microsoft.DesktopAppInstaller_1.17.10271.0_x64__8wekyb3d8bbwe\winget.exe" install --id Citrix.Workspace --silent --accept-package-agreements --accept-source-agreements --override "/silent /noreboot /includeSSON /forceinstall"`

And the package is downloaded and installed correctly.

But, now (on my machine) there's a problem with winget-install.ps1 not continuing...
winget.exe ends with:
```
Starting package install...
Successfully installed
```
The process is gone (Start-Process was started with **-Wait**) so it should continue, but it just halted there....
...C:\ProgramData\Winget-AutoUpdate\logs\install.log:
```
21:56:24 - ###   2022-04-12 - NEW INSTALL REQUEST   ###
21:56:24 - Using following Winget Cmd: C:\Program Files\WindowsApps\Microsoft.DesktopAppInstaller_1.17.10271.0_x64__8wekyb3d8bbwe\winget.exe
21:56:40 - Citrix.Workspace exists on Winget Repository.
21:56:41 - Installing Citrix.Workspace...
```
Why, I don't know!!!

**Reply by KnifMelti** [reply]
No problem installing 7-Zip (7zip.7zip) with **Start-Process -Wait**:
`powershell.exe -Executionpolicy Bypass -File winget-install.ps1 -AppIDs 7zip.7zip`

winget-install.ps1 continues as it should...

**Reply by KnifMelti** [reply]
Solution: `Get-Process winget -ErrorAction SilentlyContinue | Foreach-Object { $_.WaitForExit() }`

```
#Install App
Write-Log "Installing $AppID..." "Yellow"
$Command = "install --id $AppID --silent --accept-package-agreements --accept-source-agreements $AppArgs"
Start-Process -FilePath "$winget" -ArgumentList "$Command" -NoNewWindow
Get-Process winget -ErrorAction SilentlyContinue | Foreach-Object { $_.WaitForExit() }
```
```
#Uninstall App
Write-Log "Uninstalling $AppID..." "Yellow"
$Command = "uninstall --id $AppID --silent --accept-source-agreements $AppArgs"
Start-Process -FilePath "$winget" -ArgumentList "$Command" -NoNewWindow
Get-Process winget -ErrorAction SilentlyContinue | Foreach-Object { $_.WaitForExit() }
```

**Reply by Romanitho** [reply]
Weird that `-wait` doesn't not work as expected

**Reply by KnifMelti** [reply]
Yes, but when reading on https://docs.microsoft.com/en-us/powershell/module/microsoft.powershell.management/start-process?view=powershell-7.2
It states:
"-Wait
Indicates that this cmdlet waits for the specified process **and its descendants** to complete before accepting more input. This parameter suppresses the command prompt or **retains the window until the processes finish**."

I reckon that some process is started in the installation that never ends (or something)...

**Reply by Romanitho** [reply]
I see. Your workaround seems good anyway ;)

**Comment by KnifMelti** [comment]
#21

**Comment by KnifMelti** [comment, marked as answer]
@sschmocker did our answers help you anything?
Release 1.4.1 has the Start-Process/Get-Process implementation and Citrix Workspace can be installed with:
`powershell.exe -Executionpolicy Bypass -File winget-install.ps1 -AppIDs "Citrix.Workspace --override \"/silent /noreboot /includeSSON /forceinstall\""`

@Romanitho the $AppArgs should be removed from winget-install.ps1 when uninstalling (but maybe --override gets implemented in the future?). 

And, as a security measure:
If you're running this in system context, you should start powershell/winget-install.ps1 with **-Windowstyle** hidden!
Otherwise the user can run  a browser via the powershell window properties (Microsoft provides the links...) and in the context of System do anything on the client!

**Reply by Romanitho** [reply]
I agree, we could remove it for the uninstall command. And if MS implement the feature, it will be easy to put it back again.

Regarding system context, as a user you shouldn't see the powershell window (this is why you should not see the powershell run by the scheduled task)

**Reply by Romanitho** [reply]
And if you run in user context with `windowstyle hidden`, you might observe a flash PowerShell window (quickly hide after run). To avoid this, we use the little vbs script for the notification script for example (WAU).

**Reply by KnifMelti** [reply]
@Romanitho Your absolutely right about system context! My brain was fried in the morning ;)
